File: PharmacyProjectBack/src/main/java/com/example/demo/service/MedicineService.java
package com.example.demo.service;

import java.util.ArrayList;
import java.util.List;

import com.example.demo.dto.MedicineDTO;
import com.example.demo.dto.MedicineReservationDTO;
import com.example.demo.model.Medicine;
import com.example.demo.model.MedicineReservation;


public interface MedicineService {
     List<Medicine> findAllMedicines();
    Medicine addMedicine(MedicineDTO medicine);
    Medicine findById(Long id);
    List<Medicine> addAlternatives(ArrayList<String> alternativesId,String id);
    List<Medicine> searchMedicines(String name);
    List<MedicineDTO> filterMedicines(List<MedicineDTO> medicines,String type);
    MedicineReservationDTO makeReservation(MedicineReservationDTO med,String patientId);
    List<MedicineReservationDTO> getReservationsByPatient(Long patientId);
    MedicineReservation cancelReservation(MedicineReservationDTO med);
}

File: PharmacyProjectFront/src/app/service/medicine.service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { Medicine } from "../model/medicine";
import { MedicineForReservation } from "../model/medicineForReservation";
import { Pharmacy } from "../model/pharmacy";
import { SearchPharmacy } from "../model/searchPharmacy";

@Injectable()
export class MedicineService {
  constructor(private http: HttpClient) {
  }

  public getAllMedicines(): Observable<Medicine[]> {
    return this.http.get<Medicine[]>("http://localhost:8081/api/medicines/getAllMedicines");
  }
  public searchMedicines(sp:String): Observable<Medicine[]> {
    return this.http.post<Medicine[]>("http://localhost:8081/api/medicines/searchMedicines",sp);
  }
  public filterMedicines(sp:String,allMedicines:Medicine[]): Observable<Medicine[]> {
    return this.http.post<Medicine[]>("http://localhost:8081/api/medicines/filterMedicines/"+sp,allMedicines);
  }
  public makeReservation(med:MedicineForReservation,patientId:string): Observable<MedicineForReservation[]> {
    return this.http.post<MedicineForReservation[]>("http://localhost:8081/api/medicines/makeReservation/"+patientId,med);
  }
  public getReservationsByPatient(patientId:string): Observable<MedicineForReservation[]> {
    return this.http.get<MedicineForReservation[]>("http://localhost:8081/api/medicines/getReservationsByPatient/"+patientId);
  }
  public cancelReservation(reservation:MedicineForReservation) :Observable<Object>{
    return this.http.post<Object>("http://localhost:8081/api/medicines/cancelReservation",reservation);
  }
}
